students/nDruP/lesson03/strformat_lab.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+
+def format_quadruple(quadruple):
+    """
+    The first element is used to generate a filename that can help with file sorting.
+    The second element is a floating point number. You should display it with 2 decimal places shown.
+    The third value is an integer, but could be any number. You should display it in scientific notation, with 2 decimal places shown.
+    The fourth value is a float with a lot of digits – display it in scientific notation with 3 significant figures.
+    """
+    return "file_{0[0]:03} :   {0[1]:.2f}, {0[2]:.2e}, {0[3]:.2e}".format(quadruple)
+
+def format_quadruple_alt(quadruple):
+    """
+    Use an alternate type of format string to replicate results of format_quadruple. (keywords/f-strings)
+    """
+    prec = .2 
+    return f"file_{quadruple[0]:03} :   {quadruple[1]:{prec}f}, {quadruple[2]:{prec}e}, {quadruple[3]:{prec}e}"
+
+def format_num_string(num_tuple):
+    """
+    Rewrite: "the 3 numbers are: {:d}, {:d}, {:d}".format(1,2,3)
+    to take an arbitrary number of values.
+    """
+    if num_tuple:
+        disp_string = "The "+str(len(num_tuple))+f" numbers are: "
+        for x in range(len(num_tuple)-1):
+            disp_string+="{:d}"+f"{',' if len(num_tuple)>2 else ''} "
+        disp_string+="and {:d}"
+        return disp_string.format(*num_tuple)
+    return "There are no numbers given"
+
+def format_quintuple(quintuple):
+    """
+    Take quintuple (a,b,c,d,e) and return 'd e c a b' padded w/ a zero if any of the values are single-digits
+    """
+    return "{3:02d} {4:02d} {2:02d} {0:02d} {1:02d}".format(*quintuple)
+
+def format_fruit_weights(f_list):
+    """
+    Given the 4-element list: ['oranges',1.3,'lemons',1.1]
+    Write an f-string that will display: The weight of an orange is 1.3 and the weight of a lemon is 1.1
+    Then see if you can change the f-string so that is displays the names of the fruit in upper case, and the weight 20% higher.
+    """
+    fruits_weights = ['','']
+    for x in [0,2]:
+        fruit = (f_list[x].lower()).capitalize()
+        if fruit[len(fruit)-1].lower()=='s':
+            fruit = fruit[:len(fruit)-1]
+        weight = f_list[x+1]
+        weight_fruit = f"weight of a{'n' if fruit[0] in ['A','E','I','O','U'] else ''} {fruit} is {1.2*weight}"
+        fruits_weights[x-x//2] = weight_fruit
+    asdf = "The "+fruits_weights[0]+" and the "+fruits_weights[1]
+    return asdf
```

Comment:
I wanted to provide an alternate way to implement this function, let me know if you have any questions
```

def format_fruit_weights(f_list):
    list_of_strings = []

    fruits = f_list[::2]
    weights = f_list[1::2]

    for idx, fruit in enumerate(fruits):

        if fruit.lower().endswith("s"):
            fruit = fruit[:-1]

        fruit = fruit.lower().capitalize()
        weight = weights[idx]

        list_of_strings.append(f"weight of a{'n' if fruit[0] in ['A','E','I','O','U'] else ''} {fruit} is {1.2*weight}")

    return f"The {list_of_strings[0]} and the {list_of_strings[1]}"```